Record a new sound from the mic

paulkaplan opened the pull request:
### Resolves

_What Github issue does this resolve (please include link)?_
Fixes https://github.com/LLK/scratch-gui/issues/296

### Proposed Changes

_Describe what this Pull Request does_
Adds a new microphone recording flow:
![mic-full-flow](https://user-images.githubusercontent.com/654102/27451639-becd3492-575d-11e7-9a32-f8840c61a0b6.gif)


/cc @carljbowman @thisandagain

ericrosenbaum commented on src/lib/audio/audio-recorder.js:
```diff
@@ -0,0 +1,165 @@
+    this.mediaStreamSource.connect(this.sourceNode);
+    this.sourceNode.connect(this.analyserNode);
+    this.analyserNode.connect(this.scriptProcessorNode);
+    this.scriptProcessorNode.connect(this.audioContext.destination);
```
This line is unnecessary as far as I know!

paulkaplan replied:
(addressed offline, for posterity) it looks like it is needed.

paulkaplan commented:
I'm wondering if the `lib/audio` code should really live in a separate repo (perhaps scratch-audio?) At first I thought the functionality was completely separate from scratch-audio, but since seeing @ericrosenbaums work recently, there is definitely a lot of overlap (like sampling RMS from an `analyserNode` and playing sound buffers). It would also keep the GUI as just front-end components. @ericrosenbaum what do you think about that? Is there a way we could use the `audioEngine` for the visualization / recording / playback I'm doing here?

mrjacobbloom commented:
I'd lump the audio editor in with the paint editor, which seems big enough to live in its own repo. Just my 2 cents

rschamp commented on src/containers/record-modal.jsx:
```diff
@@ -0,0 +1,123 @@
+                storage.builtinHelper.cache(
+                    storage.AssetType.Sound,
+                    storage.DataFormat.WAV,
+                    new Uint8Array(wavBuffer), // @todo ask cwf about the need for this?
```
@cwillisf There's a question here for you ^

rschamp commented on src/containers/record-modal.jsx:
```diff
@@ -0,0 +1,123 @@
+    handleSubmit () {
+        this.setState({encoding: true});
+
+        // Allow enough time for UI refresh before starting encoding
```
What's the refresh we're waiting for? Can this be scheduled more accurately than with a timeout?

paulkaplan replied:
So my hope was that if the wav encoding took a bit of time, we would have enough time to switch this to say `Loading...` before starting the encoding process. Does putting this in a callback of `setState` do that?

rschamp replied:
Oh, I see. Yeah, as far as I know, the callback works for that purpose. I think it's called after the re-render. `componentDidUpdate` is the recommended way, but that doesn't seem right for a handler like this.

rschamp commented on src/lib/audio/audio-buffer-player.js:
```diff
@@ -0,0 +1,26 @@
+const SharedAudioContext = require('./shared-audio-context.js');
+
+const AudioBufferPlayer = function (samples) {
```
Is there a reason this doesn't use the new class syntax?

rschamp commented on src/lib/audio/audio-recorder.js:
```diff
@@ -0,0 +1,128 @@
+const SharedAudioContext = require('./shared-audio-context.js');
+
+const AudioRecorder = function () {
```
Same question here re: class syntax.

paulkaplan commented on src/components/record-modal/record-modal.css:
```diff
@@ -0,0 +1,183 @@
+    bottom: 0;
+    z-index: 1000;
+    background-color: rgba(94, 102, 125, .95);
+}
```
Yeah we no longer have a generic small modal component (the large modal component is now full-screen and uses a text filter in the nav bar, so isn't general enough anymore). Should probably be refactored.

rschamp commented:
We should probably resolve the way to access the modal before merging.

paulkaplan commented:
Almost forgot about that ray. I added that and also refactored the way the modal worked so that it could be used by the three modal-y things (prompt, recorder and library). Paired with Carl on the add buttons. There are some unusual border-related edge cases, but we aren't concerned about them. 

![image](https://user-images.githubusercontent.com/654102/27592817-6d0eb288-5b23-11e7-9fa0-d74b64166e5d.png)
